Parse existing tripleo-heat-template to create a new tht and ansible role for ansible-based TripleO deployment

elfiesmelfie opened the pull request:
See README for details.

Parses the sections of an existing heat template and populates a tripleo_<service> role and a <service>-container-ansible.yaml heat template. 
This doesn't automate everything, but instructions are given on what steps should be taken to update the outputted files.

This tool assumes that you  have an ansible role to configure your service, and that you will add this role to the appropriate deploy step in tripleo_ansible.

elfiesmelfie committed:
Start dev on converting existing THT template to -ansible

Added in the option to parse from an existing THT template
i.e. <myservice>-container-puppet.yaml
It currently populates the parameters section

There are more TODOs, documented in README.rst.

elfiesmelfie committed:
Automatically parse the output location for tht from the tht_input var

elfiesmelfie committed:
Fix indentation issue in tht template

elfiesmelfie committed:
Add in the kolla_config, docker_config and other sections from original tht

elfiesmelfie committed:
Get the deploy stages and puppet_config from the input file

elfiesmelfie committed:
Create the <deploy_stage>.yml tasks from existing sections in THT

elfiesmelfie committed:
Add block in tripleo deploy tasks template to choose new/existing tht

elfiesmelfie committed:
Add in the param renaming/merging with MyServiceVars

leifmadsen commented on README.rst:
```diff
@@ -24,3 +24,23 @@ You can override these by modifying the ``default.inv`` file, or by passing
+
+To try out this feature, pass the ``tht_input`` extra parameter to ansible::
+
+   ansible-playbook -i defauly.inv start_ansibilisation.yml -e tht_input=/path/to/existing/template
```
```suggestion
   ansible-playbook -i default.inv start_ansibilisation.yml -e tht_input=/path/to/existing/template
```

csibbitt commented on templates/THT/myservice-container-ansible.yaml.j2:
```diff
@@ -66,42 +36,61 @@ parameters:
-        #######################################################################
+        {{ tht.outputs.role_data.value.docker_config | to_nice_yaml(width=72, indent=2) | indent(width=8, indentfirst=false) }}
+
+## TODO(efoley): parse the tht.outputs.role_data.value.*_tasks and make a list of deploy_stages
```
Just to make sure I understand - this TODO is about dynamically parsing the THT to find the specific deploy_stages that might apply to this service, as they may differ from the static list found in in _tasks/main.yml_?

elfiesmelfie replied:
This one is for me, some of the others are meant to be rendered and are notes for the user

elfiesmelfie replied:
Actually, this comment is not needed, since I built a list of deploy_stages from the inputted THT

csibbitt commented on templates/tripleo_myservice/tasks/main.yml.j2:
```diff
@@ -19,4 +19,4 @@
 
 - name: Include tasks for deploy stage
-  include_tasks: {{ "{{ deploy_stage }}.yml" }}
+  include_tasks: {{ '"{{' }} deploy_stage {{ '}}' }}.yml"
```
I think this is meant to be `include_tasks: "{{ deploy_stage }}.yaml"` but allow for spaces in the deploy_stage?

Does `include_tasks: "\"{{ deploy_stage }}.yaml\""` work?

elfiesmelfie replied:
deploy_stage should not have spaces, it should be one of a number of values (host_prep_tasks, upgrade_tasks, external_deploy_tasks, etc), as listed in [1].

The strangeness with the `{{` and the `"` is because the final output will literally be `"{{ deploy_stage }}.yml"` since the entry point for this will always be main.yml, and the deploy_stage var will choose the task to import once the fact setting is done.
See [2] for example.

[1] https://docs.openstack.org/tripleo-docs/latest/developer/tht_walkthrough/service_template_sections.html
[2] https://github.com/infrawatch/tripleo-collectd-ansible-role/blob/master/tasks/main.yml

csibbitt replied:
Okay, I think I'll have to play with this myself to truly understand, but if it's working, that's a great starting point! :) Thanks.

csibbitt replied:
I re-read this today after having more thoroughly followed the code yesterday and this makes perfect sense now. Thanks! :)

elfiesmelfie replied:
Creating a template where the output also has templating mark-up is a bit of a mindfork.

csibbitt reviewed (commented):
Reading through this mostly makes sense to me. I'm going to try it out on the qdr service and play around with the output to make sure I understand what it's all for. Will be back with an "Approve" if it works out. :)

elfiesmelfie committed:
Update README.rst

Co-authored-by: Leif Madsen <lmadsen@redhat.com>

elfiesmelfie committed:
Remove TODO and update deploy_stages default list.

elfiesmelfie committed:
Add in note about defaults in tripleo_myservice

elfiesmelfie committed:
update README.rst

elfiesmelfie committed:
Add template to populate the defaults/main for tripleo_myservice

The parameters are taken from the existing THT and their
description/default values are included

elfiesmelfie commented on templates/tripleo_myservice/defaults/main.yml.j2:
```diff
@@ -0,0 +1,28 @@
+{%- import 'templates/macros.yml' as macros -%}
```
FYI, you can validate this with the first molecule job. 
I suggest that the first molecule test should run the task that imports the external <service>-config-ansible-role.

elfiesmelfie commented on templates/THT/myservice-container-ansible.yaml.j2:
```diff
@@ -66,42 +36,59 @@ parameters:
+          # The values here will be available in the tripleo_{{ myservice }}
+          # (and the {{ myservice }} ansible role.
+          #####################################################################
+          # TODO: Update names as appropriate, consider using tripleo_* as a
```
I no longer recommend updating anything here. It's unintuitive for the operator using the new THT templates. It requires knowing about how the vars are used under the hood, which is not something that the operator should care about.  Any updates to here can be made if/when particular params are deprecated, and appropriate notice will be given to the operator on how to replace those options.

In the tripleo_myservice ansible role, these can be re-mapped, if appropriate.

csibbitt reviewed (approved):
I've run this a few times now and am starting to get comfortable with it's outputs. I thing we should merge this and I will contribute back if I run into any hiccups or invent improvements.
